clients/src/main/java/org/apache/kafka/common/protocol/RecordsWriter.java

```diff
@@ -0,0 +1,141 @@
+ *
+ * For example,
+ *
+ * <pre>
+ *     recordsWritable.writeInt(10);
+ *     recordsWritable.writeRecords(records1);
+ *     recordsWritable.writeInt(20);
+ *     recordsWritable.writeRecords(records2);
+ *     recordsWritable.writeInt(30);
+ *     recordsWritable.flush();
+ * </pre>
+ *
+ * Will pass 5 Send objects to the consumer given in the constructor. Care must be taken by callers to flush any
+ * pending bytes at the end of the writing sequence to ensure everything is flushed to the consumer. This class is
+ * intended to be used with {@link org.apache.kafka.common.record.MultiRecordsSend}.
+ *
+ * @see org.apache.kafka.common.requests.FetchResponse
+ */
+public class RecordsWriter implements Writable {
+    private final String dest;
+    private final Consumer<Send> sendConsumer;
+    private final ByteArrayOutputStream byteArrayOutputStream;
+    private final DataOutput output;
+
+    public RecordsWriter(String dest, Consumer<Send> sendConsumer) {
+        this.dest = dest;
+        this.sendConsumer = sendConsumer;
+        this.byteArrayOutputStream = new ByteArrayOutputStream();
+        this.output = new DataOutputStream(this.byteArrayOutputStream);
+    }
+
+    @Override
+    public void writeByte(byte val) {
+        writeQuietly(() -> output.writeByte(val));
+    }
+
+    @Override
+    public void writeShort(short val) {
+        writeQuietly(() -> output.writeShort(val));
+    }
+
+    @Override
+    public void writeInt(int val) {
+        writeQuietly(() -> output.writeInt(val));
+    }
+
+    @Override
+    public void writeLong(long val) {
+        writeQuietly(() -> output.writeLong(val));
+
+    }
+
+    @Override
+    public void writeDouble(double val) {
+        writeQuietly(() -> ByteUtils.writeDouble(val, output));
+
+    }
+
+    @Override
+    public void writeByteArray(byte[] arr) {
+        writeQuietly(() -> output.write(arr));
+    }
+
+    @Override
+    public void writeUnsignedVarint(int i) {
+        writeQuietly(() -> ByteUtils.writeUnsignedVarint(i, output));
+    }
+
+    @Override
+    public void writeByteBuffer(ByteBuffer src) {
+        writeQuietly(() -> output.write(src.array(), src.position(), src.remaining()));
+    }
+
+    @FunctionalInterface
+    private interface IOExceptionThrowingRunnable {
+        void run() throws IOException;
+    }
+
+    private void writeQuietly(IOExceptionThrowingRunnable runnable) {
+        try {
+            runnable.run();
+        } catch (IOException e) {
+            throw new RuntimeException("Writable encountered an IO error", e);
+        }
+    }
+
+    @Override
+    public void writeRecords(BaseRecords records) {
+        flush();
+        sendConsumer.accept(records.toSend(dest));
+    }
+
+    /**
+     * Flush any pending bytes as a ByteBufferSend and reset the buffer
+     */
+    public void flush() {
+        ByteBufferSend send = new ByteBufferSend(dest,
+                ByteBuffer.wrap(byteArrayOutputStream.toByteArray()));
```

Comment:
This creates a copy of the underlying bytes, can we avoid it?

Reply:
Yea, it's possible, but rather complicated I think. We would need to manage our own byte array and grow it on-demand (like what happens in ByteArrayOutputStream). Then we could use `ByteBuffer#slice` to pass views of this array to the ByteBufferSend objects. I don't think this current approach is any worse than before in terms of array allocations, so maybe we could save this for a future optimization?

Reply:
Would `ByteBufferOutputStream` be useful here?

Reply:
Looks like the expansion factor for ByteArrayOutputStream varies on the JDK version. In [JDK 8](https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/java/io/ByteArrayOutputStream.java#L105) and [11](https://github.com/openjdk/jdk/blob/jdk-11+28/src/java.base/share/classes/java/io/ByteArrayOutputStream.java) it's 2x, but in [JDK 14](https://github.com/openjdk/jdk/blob/jdk-14+36/src/java.base/share/classes/java/io/ByteArrayOutputStream.java#L95) it just grows the buffer to the minimum needed size.

Our growth factor of 1.1 in ByteBufferOutputStream seems reasonable . Not to mention avoiding the final copy by using slice would be nice too.

Reply:
@mumrah Thanks for checking this. However, the behavior in JDK 14 has not changed in that way. Performance would be atrocious if it did:

```java
private void ensureCapacity(int minCapacity) {
        // overflow-conscious code
        int oldCapacity = buf.length;
        int minGrowth = minCapacity - oldCapacity;
        if (minGrowth > 0) {
            buf = Arrays.copyOf(buf, ArraysSupport.newLength(oldCapacity,
                    minGrowth, oldCapacity /* preferred growth */));
        }
```

The third parameter passed to `newLength` is the preferred growth, which is `oldCapacity`. That is, it doubles if it doesn't cause overflow. We should probably double for `ByteBufferOutputStream` too _if_ we have no estimate of the expected size. `1.1` makes sense if we do have an estimate (which is the case in current usage, I believe).

Reply:
Looking a bit more, it seems like this will be mostly used by the data that precedes the actual records. Do we have a sense for what's the typical size for that? If we do, we can use that in the initial size and we can keep the `1.1` growth.

Reply:
Thanks for the explanation, @ijuma. I missed the semantics of `newLength`

After the initial few top-level fields, each partition will have something like 38 bytes preceding its records. Maybe we could increase initial capacity to 64 bytes?

Reply:
Sounds good.

Reply:
I increased the initial buffer size to 64 and also added 2x growth factor for the buffer. It occurred to me the initial size only really helps for the first partition's header fields, but beyond that (since we are reusing/growing the same ByteBufferOutputStream) we don't know what we'll need. The JMH benchmark did confirm that 2x was more performant than 1.1x for FetchResponse.

Existing usages of ByteBufferOutputStream were not modified and still use 1.1x